File: backend/internal/awsx/grading.go
package awsx

// The grader's view of this environment. The keys are the ones the 2026
// national task-3 scoring sheet (and its results_<비번호>.log) actually carry:
// "image download" and "Exception Handling" (비정상 요청 처리), and per-API
// "availability" and "performance ≤ SLO" for user · product · stress. Each key
// is measured from the source that can see it — the app log for the four served
// ratios, the WAF log for the 403 side of Exception Handling.
//
// The scoring itself is pure; the AWS-touching function runs the Insights
// queries and hands rows to the builder. No points are assigned anywhere here —
// the score is the grader's to compute from its own run.

import (
	"context"
	"fmt"
	"math"
	"strconv"
	"strings"

	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/analysis"
	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/config"
	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/types"
)

// The task sheet: every API must answer within 5 s to count as served at all;
// user and product aim at 0.2 s, stress at 1 s.
const AvailDeadlineMs = 5_000

var SloMs = map[string]int{"user": 200, "product": 200, "stress": 1_000}

var GradingAPIs = []string{"user", "product", "stress"}

// measure: observed value for one grading key.
func measure(label string, okCount, total int, source string, approximate bool) types.GradingScore {
	pct := 0.0
	if total > 0 {
		pct = math.Round(float64(okCount)/float64(total)*1000) / 10
	}
	return types.GradingScore{Label: label, Pct: pct, OkCount: okCount, Total: total, Source: source, Approximate: approximate}
}

// trimTrailingSlash: "/v1/user/" and "/v1/user" are one route; only the
// comparison needs to say so. "/" itself is left alone — it is a path, not a
// trailing slash.
func trimTrailingSlash(path string) string {
	if len(path) > 1 {
		return strings.TrimSuffix(path, "/")
	}
	return path
}

// APIOf says which API a path belongs to; "" when the path is off the served
// surface and never counts toward availability.
func APIOf(path string) string {
	p := path
	if i := strings.IndexByte(p, '?'); i >= 0 {
		p = p[:i]
	}
	for _, api := range GradingAPIs {
		base := "/v1/" + api
		if p == base || strings.HasPrefix(p, base+"/") {
			return api
		}
	}
	return ""
}

// IsImagePath: image delivery is S3 objects served under /images/<object path>
// through the same endpoint as the APIs. It is a scoring key of its own ("image
// download"), not part of any API's availability, so it has to be split out
// before APIOf ever sees the row. Anchored at a segment boundary so a route
// that merely starts with the same letters (/imagesearch) does not collide.
func IsImagePath(path string) bool {
	p := path
	if i := strings.IndexByte(p, '?'); i >= 0 {
		p = p[:i]
	}
	return p == "/images" || strings.HasPrefix(p, "/images/")
}

// BuildGradingQuery counts availability, both SLO tiers and the contract
// codes in the query rather than pulling rows back — the rows would be the
// whole traffic volume. Grouped by route (no query string), or every GET is
// its own row because the grader's requestid rides in the query.
func BuildGradingQuery() string {
	return strings.Join([]string{
		analysis.ParseFields,
		analysis.AccessLogFilter,
		"stats count(*) as total," +
			fmt.Sprintf(" sum(status >= 200 and status < 300 and latency_ms <= %d) as availOk,", AvailDeadlineMs) +
			fmt.Sprintf(" sum(status >= 200 and status < 300 and latency_ms <= %d) as fastOk,", SloMs["user"]) +
			fmt.Sprintf(" sum(status >= 200 and status < 300 and latency_ms <= %d) as slowOk,", SloMs["stress"]) +
			" sum(status = 404) as notFound," +
			" sum(status = 403) as forbidden," +
			" sum(status >= 500) as serverErr" +
			" by path",
		"sort total desc",
		"limit 200",
	}, " | ")
}

// BuildTrapQuery counts the product binary's trap line — one per Attacker-Bot
// request that reached the app instead of being blocked.
func BuildTrapQuery() string {
	return fmt.Sprintf(`filter @message like "%s" | stats count(*) as leaked`, analysis.MaliciousTrapLine)
}

// BuildWafGradingQuery folds the WAF log by uri × method × action. uri in the
// WAF log is the path without its query string, so it groups like `path` does
// on the app side.
func BuildWafGradingQuery() string {
	return "stats count(*) as cnt by httpRequest.uri as uri, httpRequest.httpMethod as method, action | sort cnt desc | limit 500"
}

type PathRow struct {
	Path      string
	Total     int
	AvailOk   int
	FastOk    int
	SlowOk    int
	NotFound  int
	Forbidden int
	ServerErr int
}

func rowInt(row InsightsRow, k string) int {
	n, err := strconv.ParseFloat(row[k], 64)
	if err != nil {
		return 0
	}
	return int(n)
}

func ToPathRow(row InsightsRow) PathRow {
	return PathRow{
		Path:      analysis.CleanPath(row["path"]),
		Total:     rowInt(row, "total"),
		AvailOk:   rowInt(row, "availOk"),
		FastOk:    rowInt(row, "fastOk"),
		SlowOk:    rowInt(row, "slowOk"),
		NotFound:  rowInt(row, "notFound"),
		Forbidden: rowInt(row, "forbidden"),
		ServerErr: rowInt(row, "serverErr"),
	}
}

// WafRow is one uri × method × action cell of the WAF log.
type WafRow struct {
	URI    string
	Method string
	Action string
	Count  int
}

func ToWafRow(row InsightsRow) WafRow {
	return WafRow{
		URI:    row["uri"],
		Method: strings.ToUpper(row["method"]),
		Action: strings.ToUpper(row["action"]),
		Count:  rowInt(row, "cnt"),
	}
}

type GradingParams struct {
	Rows []PathRow
	// The WAF log fold; nil when no WAF log group is configured (WafAvailable
	// false). Requests the firewall blocked never reach the app, so they are not
	// in Rows — they are what the 403 keys are made of.
	WafRows      []WafRow
	WafAvailable bool
	// Attacker-Bot requests the app served (and answered 500) — abnormal
	// requests that leaked past the WAF.
	TrapLeaked   int
	Window       types.ResolvedWindow
	Source       string
	ScannedBytes int64
	Notes        []string
}

// scoreBandsAvailability: the score bands the 2026 sheet pays on, highest
// first. Every percentage key uses the same ladder, and each rung crossed is
// worth 0.5점 — so the number an operator needs is not "86.2%" but "one band
// below the next 0.5, and 1.3%p away from it". Kept as numbers rather than a
// formatted string so the gap can be computed against them.
var scoreBandsAvailability = []float64{90, 87.5, 85, 82.5, 80, 70, 50, 30}

// scoreBandsAbnormal: 비정상 요청 처리 (image download · Exception Handling)
// pays on four rungs, not eight. Same shape, different ladder — do not merge
// them.
var scoreBandsAbnormal = []float64{90, 85, 80, 50}

// trimFloat formats a band or gap the way a JS template literal would: 90 → "90",
// 87.5 → "87.5", 4 → "4" — no trailing zeros, so the console reads clean.
func trimFloat(v float64) string {
	return strconv.FormatFloat(v, 'f', -1, 64)
}

// bandOf returns the band a percentage currently sits in and the next one up.
func bandOf(pct float64, bands []float64) (tier, nextTier *string) {
	// bands is descending, so the first rung at or below the value is the one it
	// has already earned; the rung above it is what the next 0.5점 costs.
	earnedIndex := -1
	for i, b := range bands {
		if pct >= b {
			earnedIndex = i
			break
		}
	}
	if earnedIndex >= 0 {
		tier = types.Ptr(fmt.Sprintf("%s%% 구간", trimFloat(bands[earnedIndex])))
	}
	var nextValue float64
	haveNext := false
	if earnedIndex == -1 {
		nextValue = bands[len(bands)-1]
		haveNext = true
	} else if earnedIndex-1 >= 0 {
		nextValue = bands[earnedIndex-1]
		haveNext = true
	}
	if !haveNext {
		return tier, nil
	}
	gap := math.Round((nextValue-pct)*10) / 10
	nextTier = types.Ptr(fmt.Sprintf("%s%% 까지 %s%%p", trimFloat(nextValue), trimFloat(gap)))
	return tier, nextTier
}

// BuildGradingPanel builds the panel from already-fetched pieces so the
// scoring itself is pure.
func BuildGradingPanel(p GradingParams) types.GradingPanel {
	type apiAcc struct{ total, availOk, sloOk int }
	perAPI := map[string]*apiAcc{}
	for _, api := range GradingAPIs {
		perAPI[api] = &apiAcc{}
	}

	// Image delivery is its own scoring key ("image download") and its own
	// surface: S3 objects served under /images/ through the same endpoint. Its
	// SLO is the availability deadline itself — 5s for both — so there is no
	// second tier.
	imageTotal, imageOk := 0, 0

	// Requests to paths the task does not serve. The contract is 404 there, and
	// "Exception Handling" is what the grader calls the ratio that ends
	// correctly.
	undefTotal, undefOk := 0, 0

	for _, r := range p.Rows {
		if IsImagePath(r.Path) {
			imageTotal += r.Total
			imageOk += r.AvailOk
			continue
		}
		api := APIOf(r.Path)
		if api == "" {
			if !config.IsLowPriorityPath(r.Path) {
				undefTotal += r.Total
				undefOk += r.NotFound
			}
			continue
		}
		acc := perAPI[api]
		acc.total += r.Total
		acc.availOk += r.AvailOk
		// The SLO differs per API, so the right column is picked here rather
		// than baked into the query.
		if SloMs[api] <= SloMs["user"] {
			acc.sloOk += r.FastOk
		} else {
			acc.sloOk += r.SlowOk
		}
	}

	// WAF side. A block on the served surface is an abnormal request answered
	// 403 — which is what the task asks for. A block on an undefined path is a
	// 404 that became a 403, which is the violation, so it only ever enlarges
	// the denominator. A block on the image surface is neither — it would cost
	// image download points — so it is excluded before either bucket.
	wafServedBlocked, wafUndefBlocked := 0, 0
	for _, w := range p.WafRows {
		if w.Action != "BLOCK" {
			continue
		}
		if IsImagePath(w.URI) {
			continue
		}
		if APIOf(w.URI) != "" {
			wafServedBlocked += w.Count
			continue
		}
		if !config.IsLowPriorityPath(w.URI) {
			wafUndefBlocked += w.Count
		}
	}

	const appSrc = "앱 로그"
	wafSrc := "WAF 로그"
	if !p.WafAvailable {
		wafSrc = "WAF 로그 없음"
	}
	lines := []types.GradingScore{}

	push := func(label string, okCount, total int, source string, bands []float64, approximate bool) {
		line := measure(label, okCount, total, source, approximate)
		// A band is a claim about data; when nothing was observed there is no
		// claim to make, so both stay nil.
		if total > 0 {
			line.Tier, line.NextTier = bandOf(line.Pct, bands)
		}
		lines = append(lines, line)
	}

	// Ordered exactly as the sheet lists them, so the two read side by side.
	push("image download", imageOk, imageTotal, appSrc, scoreBandsAbnormal, false)
	// Numerator = abnormal requests the WAF turned into 403 + undefined-path
	// requests the app ended as 404. Denominator adds what leaked to the app and
	// what the WAF wrongly blocked on an undefined path.
	push(
		"Exception Handling",
		wafServedBlocked+undefOk,
		wafServedBlocked+undefOk+p.TrapLeaked+(undefTotal-undefOk)+wafUndefBlocked,
		fmt.Sprintf("%s + %s", wafSrc, appSrc),
		scoreBandsAbnormal,
		true,
	)
	for _, api := range GradingAPIs {
		a := perAPI[api]
		push(fmt.Sprintf("(%s) availability", api), a.availOk, a.total, appSrc, scoreBandsAvailability, false)
	}
	for _, api := range GradingAPIs {
		a := perAPI[api]
		push(fmt.Sprintf("(%s) performance ≤ %.1fs", api, float64(SloMs[api])/1000), a.sloOk, a.total, appSrc, scoreBandsAvailability, false)
	}

	notes := append([]string{}, p.Notes...)
	notes = append(notes,
		"채점기 로그(results_<비번호>.log)의 키 이름을 그대로 썼다 — image download · Exception Handling · (api) availability · (api) performance. cost ratio 는 이 화면이 아니라 아래 노드 대수 패널이 다룬다.",
		fmt.Sprintf("availability = 2xx && %ds 이내. performance = 그중 SLO(user·product %dms / stress %dms) 이내. image download 는 둘 다 %ds 라 한 줄뿐이다.", AvailDeadlineMs/1000, SloMs["user"], SloMs["stress"], AvailDeadlineMs/1000),
		"채점기의 응답시간은 **클라이언트 도착 기준**이고 이 표는 앱이 기록한 처리 시간이다 — 네트워크·ALB·CloudFront 구간이 빠져 있어 항상 낙관적으로 보인다. 실제 점수는 이 값보다 낮게 나온다고 보는 편이 안전하다.",
		"분모는 앱 로그의 [GIN] 액세스 라인 전체라 앱이 스스로 내는 403(username 중복)·400·500 도 들어간다. 채점기는 자신이 보낸 요청만 세므로 값이 다를 수 있다.",
		fmt.Sprintf("Exception Handling: 분자 = WAF BLOCK(서비스 경로) + 앱이 404 로 끝낸 미지정 경로. 분모에 앱까지 새어 들어온 비정상 요청 %d건과 WAF 가 미지정 경로를 막아 403 이 된 %d건을 더했다 — 후자는 그 자체로 계약 위반이다.", p.TrapLeaked, wafUndefBlocked),
		"구간 표시는 채점표의 문턱(90 / 87.5 / 85 / 82.5 / 80 / 70 / 50 / 30%, 비정상 처리는 90 / 85 / 80 / 50%)에 맞춘 것이다. 다음 문턱까지 남은 %p 가 곧 다음 0.5점이다.",
		"점수는 매기지 않는다. 이 표는 관측값을 채점기 키에 맞춰 정렬해 둔 것이고, 점수는 채점 플랫폼이 정한다.",
		"서비스 경로: "+strings.Join(config.AppTrafficPaths(), ", "),
	)
	if !p.WafAvailable {
		notes = append(notes, "WAF_LOG_GROUP 이 비어 있어 Exception Handling 의 403 쪽 분자가 0 이다 — 앱 로그만으로는 차단 건수를 볼 수 없다. 설정에서 WAF 로그 그룹을 지정하면 채워진다.")
	}

	return types.GradingPanel{
		Lines:        lines,
		Window:       p.Window,
		Source:       p.Source,
		ScannedBytes: p.ScannedBytes,
		Notes:        notes,
	}
}

// FetchGradingPanel is the one AWS-touching function here: one app-log query
// for the per-path stats, one for the trap line, and — when a WAF log group is
// configured — one WAF-log fold.
func (a *AWS) FetchGradingPanel(ctx context.Context, win types.ResolvedWindow) (types.GradingPanel, error) {
	logGroup := a.Settings.AppLogGroup()
	res, err := a.RunInsightsQuery(ctx, InsightsParams{
		LogGroup: logGroup,
		Query:    BuildGradingQuery(),
		StartMs:  &win.StartMs,
		EndMs:    &win.EndMs,
	})
	if err != nil {
		return types.GradingPanel{}, err
	}
	rows := make([]PathRow, 0, len(res.Rows))
	for _, r := range res.Rows {
		rows = append(rows, ToPathRow(r))
	}
	params := GradingParams{
		Rows:         rows,
		Window:       win,
		Source:       fmt.Sprintf("앱 로그 Logs Insights(%s)", logGroup),
		ScannedBytes: res.BytesScanned,
		Notes:        []string{},
	}

	if trap, err := a.RunInsightsQuery(ctx, InsightsParams{
		LogGroup: logGroup, Query: BuildTrapQuery(), StartMs: &win.StartMs, EndMs: &win.EndMs,
	}); err == nil {
		params.ScannedBytes += trap.BytesScanned
		if len(trap.Rows) > 0 {
			params.TrapLeaked = rowInt(trap.Rows[0], "leaked")
		}
	} else {
		params.Notes = append(params.Notes, "Attacker-Bot trap 라인 집계 실패: "+ErrMsg(err))
	}

	if wafGroup := a.Settings.WafLogGroup(); wafGroup != "" {
		wafRes, err := a.RunInsightsQuery(ctx, InsightsParams{
			LogGroup: wafGroup,
			Region:   a.Settings.WafRegion(),
			Query:    BuildWafGradingQuery(),
			StartMs:  &win.StartMs,
			EndMs:    &win.EndMs,
		})
		if err == nil {
			params.WafAvailable = true
			params.ScannedBytes += wafRes.BytesScanned
			params.Source += fmt.Sprintf(" + WAF 로그(%s)", wafGroup)
			for _, r := range wafRes.Rows {
				params.WafRows = append(params.WafRows, ToWafRow(r))
			}
		} else {
			params.Notes = append(params.Notes, "WAF 로그 집계 실패 — 403 키는 앱 로그만으로 채움: "+ErrMsg(err))
		}
	}
	return BuildGradingPanel(params), nil
}

File: backend/awsx/grading.ts
// The grader's view of this environment. The keys follow the 2025 national
// task-3 scoring sheet (docs/binaries.md §채점 키): per-API "로드 처리"
// (availability) and "로드 처리 <= SLO" (performance) for user · product ·
// stress, "Email Request Validation" and "비정상 요청 처리율" (abnormal
// requests answered 403), plus the task's own "undefined path → 404" contract.
//
// The scoring itself is pure; the AWS-touching function runs the Insights
// queries and hands rows to the builder. No points are assigned anywhere here —
// the score is the grader's to compute from its own run.

import {
  ACCESS_LOG_FILTER,
  cleanPath,
  MALICIOUS_TRAP_LINE,
  PARSE_FIELDS,
} from "../analysis/logfields.ts";
import { appTrafficPaths, isLowPriorityPath, normalizePath } from "../config/paths.ts";
import type { GradingPanel, GradingScore, ResolvedWindow } from "../../src/lib/types.ts";
import { errMsg, type AWS } from "./clients.ts";
import { runInsightsQuery, type InsightsRow } from "./insights.ts";

/**
 * The task sheet: every API must answer within 5 s to count as served at all;
 * user and product aim at 0.2 s, stress at 1 s.
 */
export const AVAIL_DEADLINE_MS = 5_000;

export const SLO_MS: Record<string, number> = { user: 200, product: 200, stress: 1_000 };

export const GRADING_APIS = ["user", "product", "stress"];

/** Observed value for one grading key. */
function measure(
  label: string,
  okCount: number,
  total: number,
  source: string,
  approximate: boolean,
): GradingScore {
  const pct = total > 0 ? Math.round((okCount / total) * 1000) / 10 : 0;
  return { label, pct, okCount, total, source, approximate };
}

/** "/v1/user/" and "/v1/user" are one route; only the comparison needs to say so. */
function trimTrailingSlash(path: string): string {
  return path.length > 1 && path.endsWith("/") ? path.slice(0, -1) : path;
}

/**
 * Which API a path belongs to; "" when the path is off the served surface and
 * never counts toward availability.
 */
export function apiOf(path: string): string {
  const p = path.split("?")[0]!;
  for (const api of GRADING_APIS) {
    const base = "/v1/" + api;
    if (p === base || p.startsWith(base + "/")) return api;
  }
  return "";
}

/**
 * Counts availability, both SLO tiers and the contract codes in the query
 * rather than pulling rows back — the rows would be the whole traffic volume.
 * Grouped by route (no query string), or every GET is its own row because the
 * grader's requestid rides in the query.
 */
export function buildGradingQuery(): string {
  return [
    PARSE_FIELDS,
    ACCESS_LOG_FILTER,
    "stats count(*) as total," +
      ` sum(status >= 200 and status < 300 and latency_ms <= ${AVAIL_DEADLINE_MS}) as availOk,` +
      ` sum(status >= 200 and status < 300 and latency_ms <= ${SLO_MS.user}) as fastOk,` +
      ` sum(status >= 200 and status < 300 and latency_ms <= ${SLO_MS.stress}) as slowOk,` +
      " sum(status = 404) as notFound," +
      " sum(status = 403) as forbidden," +
      " sum(status >= 500) as serverErr" +
      " by path",
    "sort total desc",
    "limit 200",
  ].join(" | ");
}

/**
 * Counts the product binary's trap line — one per Attacker-Bot request that
 * reached the app instead of being blocked.
 */
export function buildTrapQuery(): string {
  return `filter @message like "${MALICIOUS_TRAP_LINE}" | stats count(*) as leaked`;
}

/**
 * Folds the WAF log by uri × method × action. uri in the WAF log is the path
 * without its query string, so it groups like `path` does on the app side.
 */
export function buildWafGradingQuery(): string {
  return "stats count(*) as cnt by httpRequest.uri as uri, httpRequest.httpMethod as method, action | sort cnt desc | limit 500";
}

export interface PathRow {
  path: string;
  total: number;
  availOk: number;
  fastOk: number;
  slowOk: number;
  notFound: number;
  forbidden: number;
  serverErr: number;
}

function rowInt(row: InsightsRow, k: string): number {
  const n = Number.parseFloat(row[k] ?? "");
  return Number.isFinite(n) ? Math.trunc(n) : 0;
}

export function toPathRow(row: InsightsRow): PathRow {
  return {
    path: cleanPath(row.path ?? ""),
    total: rowInt(row, "total"),
    availOk: rowInt(row, "availOk"),
    fastOk: rowInt(row, "fastOk"),
    slowOk: rowInt(row, "slowOk"),
    notFound: rowInt(row, "notFound"),
    forbidden: rowInt(row, "forbidden"),
    serverErr: rowInt(row, "serverErr"),
  };
}

/** One uri × method × action cell of the WAF log. */
export interface WafRow {
  uri: string;
  method: string;
  action: string;
  count: number;
}

export function toWafRow(row: InsightsRow): WafRow {
  return {
    uri: row.uri ?? "",
    method: (row.method ?? "").toUpperCase(),
    action: (row.action ?? "").toUpperCase(),
    count: rowInt(row, "cnt"),
  };
}

export interface GradingParams {
  rows: PathRow[];
  /**
   * The WAF log fold; empty when no WAF log group is configured (wafAvailable
   * false). Requests the firewall blocked never reach the app, so they are not
   * in rows — they are what the 403 keys are made of.
   */
  wafRows?: WafRow[];
  wafAvailable?: boolean;
  /**
   * Attacker-Bot requests the app served (and answered 500) — abnormal requests
   * that leaked past the WAF.
   */
  trapLeaked?: number;
  window: ResolvedWindow;
  source?: string;
  scannedBytes?: number;
  notes?: string[];
}

/** Builds the panel from already-fetched pieces so the scoring itself is pure. */
export function buildGradingPanel(p: GradingParams): GradingPanel {
  const wafRows = p.wafRows ?? [];
  const trapLeaked = p.trapLeaked ?? 0;

  const perApi = new Map<string, { total: number; availOk: number; sloOk: number }>();
  for (const api of GRADING_APIS) perApi.set(api, { total: 0, availOk: 0, sloOk: 0 });

  // Requests to paths the binaries do not serve, and how many the app ended
  // with the 404 the task requires.
  let undefTotal = 0;
  let undefOk = 0;

  for (const r of p.rows) {
    const api = apiOf(r.path);
    if (api === "") {
      if (!isLowPriorityPath(r.path)) {
        undefTotal += r.total;
        undefOk += r.notFound;
      }
      continue;
    }
    const acc = perApi.get(api)!;
    acc.total += r.total;
    acc.availOk += r.availOk;
    // The SLO differs per API, so the right column is picked here rather than
    // baked into the query.
    if (SLO_MS[api]! <= SLO_MS.user!) acc.sloOk += r.fastOk;
    else acc.sloOk += r.slowOk;
  }

  // WAF side: blocks on the served surface are the 403s the abnormal-request
  // keys count; blocks on undefined paths are 404s that became 403s.
  let wafServedBlocked = 0;
  let wafEmailBlocked = 0;
  let wafUndefBlocked = 0;
  for (const w of wafRows) {
    if (w.action !== "BLOCK") continue;
    if (apiOf(w.uri) !== "") {
      wafServedBlocked += w.count;
      // normalizePath keeps a trailing slash because NORMALIZE_PATH does, so the
      // endpoint comparison has to trim one — "/v1/user/" is the same route the
      // grader posts to, and counting it as a different one would understate the
      // email-validation key.
      if (w.method === "POST" && trimTrailingSlash(normalizePath(w.uri)) === "/v1/user") {
        wafEmailBlocked += w.count;
      }
      continue;
    }
    if (!isLowPriorityPath(w.uri)) wafUndefBlocked += w.count;
  }

  const appSrc = "앱 로그";
  const lines: GradingScore[] = [];
  for (const api of GRADING_APIS) {
    const acc = perApi.get(api)!;
    lines.push(measure(`${api} API 로드 처리`, acc.availOk, acc.total, appSrc, false));
  }
  for (const api of GRADING_APIS) {
    const acc = perApi.get(api)!;
    lines.push(
      measure(
        `${api} API 로드 처리 ≤ ${(SLO_MS[api]! / 1000).toFixed(1)}s`,
        acc.sloOk,
        acc.total,
        appSrc,
        false,
      ),
    );
  }

  const wafSrc = p.wafAvailable ? "WAF 로그" : "WAF 로그 없음";
  // Denominator = what we saw end as 403 + what we saw get through. The
  // grader's own count of what it injected is not observable anywhere.
  lines.push(
    measure(
      "Email Request Validation (403)",
      wafEmailBlocked,
      wafEmailBlocked,
      wafSrc + " · POST /v1/user 차단 건수, 분모 없음",
      true,
    ),
  );
  lines.push(
    measure(
      "비정상 요청 처리율 (403)",
      wafServedBlocked,
      wafServedBlocked + trapLeaked,
      wafSrc + " + 앱 로그 trap 라인",
      true,
    ),
  );
  lines.push(
    measure("미지정 경로 404", undefOk, undefTotal + wafUndefBlocked, appSrc + " + WAF 차단", false),
  );

  const notes = [
    ...(p.notes ?? []),
    `로드 처리 = 2xx && ${AVAIL_DEADLINE_MS / 1000}s 이내 / 해당 API 로 들어온 요청 전체. 성능 키는 그중 SLO(user·product ${SLO_MS.user}ms / stress ${SLO_MS.stress}ms) 이내.`,
    "분모는 앱 로그의 [GIN] 액세스 라인 전체라 앱이 스스로 내는 403(username 중복 → 'It already exists in a database.')·400·500 도 들어간다. 채점기는 자신이 보낸 요청만 세므로 값이 다를 수 있다.",
    `비정상 요청 처리율: 분자 = WAF BLOCK(서비스 경로), 분모 = 분자 + 앱까지 새어 들어온 Attacker-Bot 요청 ${trapLeaked}건 (product 가 'Consumed resources by malicious attacks.' 를 찍고 500 으로 응답). 채점기가 보낸 비정상 요청 전체 수는 관측 불가 — 새는 건수가 0 인지를 본다.`,
    "Email Request Validation: WAF 가 POST /v1/user 를 차단한 건수만 보인다 — 잘못된 이메일이 몇 건 주입됐는지는 어디에도 기록되지 않는다(앱은 이메일을 검사하지 않는다). 0건이면 규칙이 없거나 COUNT 상태다.",
    "미지정 경로 404: 앱 로그의 비서비스 경로 요청 중 404 로 끝난 비율. WAF 가 미지정 경로를 BLOCK 하면 403 이 나가 위반 — 그 건수는 분모에만 더했다.",
    "점수는 매기지 않는다. 이 표는 관측값을 채점기 키에 맞춰 정렬해 둔 것이고, 점수는 채점 플랫폼이 정한다.",
    "서비스 경로: " + appTrafficPaths().join(", "),
  ];
  if (!p.wafAvailable) {
    notes.push(
      "WAF_LOG_GROUP 이 비어 있어 403 키는 앱 로그만으로 채웠다 — 차단 건수는 0 으로 보인다. 설정에서 WAF 로그 그룹을 지정하면 채워진다.",
    );
  }

  return {
    lines,
    window: p.window,
    source: p.source ?? "",
    scannedBytes: p.scannedBytes ?? 0,
    notes,
  };
}

/**
 * The one AWS-touching function here: one app-log query for the per-path stats,
 * one for the trap line, and — when a WAF log group is configured — one WAF-log
 * fold.
 */
export async function fetchGradingPanel(a: AWS, win: ResolvedWindow): Promise<GradingPanel> {
  const logGroup = a.settings.appLogGroup();
  const res = await runInsightsQuery(a, {
    logGroup,
    query: buildGradingQuery(),
    startMs: win.startMs,
    endMs: win.endMs,
  });

  const params: GradingParams = {
    rows: res.rows.map(toPathRow),
    wafRows: [],
    wafAvailable: false,
    trapLeaked: 0,
    window: win,
    source: `앱 로그 Logs Insights(${logGroup})`,
    scannedBytes: res.bytesScanned,
    notes: [],
  };

  try {
    const trap = await runInsightsQuery(a, {
      logGroup,
      query: buildTrapQuery(),
      startMs: win.startMs,
      endMs: win.endMs,
    });
    params.scannedBytes = (params.scannedBytes ?? 0) + trap.bytesScanned;
    if (trap.rows.length > 0) params.trapLeaked = rowInt(trap.rows[0]!, "leaked");
  } catch (e) {
    params.notes!.push("Attacker-Bot trap 라인 집계 실패: " + errMsg(e));
  }

  const wafGroup = a.settings.wafLogGroup();
  if (wafGroup !== "") {
    try {
      const wafRes = await runInsightsQuery(a, {
        logGroup: wafGroup,
        region: a.settings.wafRegion(),
        query: buildWafGradingQuery(),
        startMs: win.startMs,
        endMs: win.endMs,
      });
      params.wafAvailable = true;
      params.scannedBytes = (params.scannedBytes ?? 0) + wafRes.bytesScanned;
      params.source += ` + WAF 로그(${wafGroup})`;
      params.wafRows = wafRes.rows.map(toWafRow);
    } catch (e) {
      params.notes!.push("WAF 로그 집계 실패 — 403 키는 앱 로그만으로 채움: " + errMsg(e));
    }
  }

  return buildGradingPanel(params);
}
